### ISSUE TITLE
Details on the GAM format

### ISSUE THREAD

**Issue by evanbiederstedt** [vgteam/vg#1688]
Hi there

I was hoping for a clarification on the GAM format. Again, I apologize if I missed this in the documentation previously:

I've converted the GAM into JSON. Let's take a look at these two:

```
{"identity": 1.0, "sequence": "G...C", "quality": "GyIiIyIjIiMiIyMjIyQiIiQkIyElJCUlJycnJickKSkpKCgpKCgoKCgnIikoKCgmKCkpKSgpKSkpKSkpKSkpKSkpKCkpKSkoKCkpKSkpKSkoKSkpKSgpKCcnJycnJSUlJSUiIiI=", 
"path": {"mapping": [{"rank": 1, "position": {"node_id": 8545, "offset": 148}, "edit": [{"from_length": 101, "to_length": 101}]}]}, 
"name": "522355372/1", "mapping_quality": 60, "score": 111, "refpos": [{"offset": 8540756, "name": "chr21"}]}

{"identity": 1.0, "sequence": "G...G", "quality": "IyMiIiIhIyMjIyMjIyQkIiQkIh4lJSUnJycnJyUoKCkoKSkpKCkpKSkpKSgoJyYpKCgmJyYpKCkoKCUnKSkoKCkoJikpJyMpKSgmKSgoJyUoJygoKSgmKScnJycnJSUlJSUiIiI=", 
"path": {"mapping": [{"rank": 1, "position": {"node_id": 8545, "is_reverse": true, "offset": 746}, "edit": [{"from_length": 101, "to_length": 101}]}]}, 
"name": "522354452/2", "mapping_quality": 60, "score": 111, "refpos": [{"is_reverse": true, "offset": 8541354, "name": "chr21"}]}
```

Certain fields are very clearly defined, e.g. sequence, mapping_quality. Others are immediately clear. 

Questions:

(1) How is the `identity` calculated? 

(2) What are the reference coordinates with respect to the graph? In this case, it appears to be 

`"refpos": [{"offset": 8540756, "name": "chr21"}]`

and

`"refpos": [{"is_reverse": true, "offset": 8541354, "name": "chr21"}]`

How does the `offset` correspond to the reference genome? It doesn't appear that I should think of this as a node, or like a start/end (as in with a linear reference)

(3) How should I interpret `quality`? 

(4) For the `path`, e.g. 

    "path": {"mapping": [{"rank": 1, "position": {"node_id": 8545, "offset": 148}, "edit": [{"from_length": 101, "to_length": 101}]}]}

What is the definition of the field `offset` here?

**Comment by adamnovak** [comment]
I can answer some of these.

For 1, `identity` is defined here:
https://github.com/vgteam/vg/blob/4460c770d66c7dacce356e511d832a31fd6d6cfc/src/vg.proto#L124
So the numerator is the exact matches, and the denominator is the "aligned" bases. Unfortunately, I don't recall how we handle softclips and indels in that denominator.

For 2, I am pretty sure that `offset` is just the linear position on the path, although I'm not sure which end of the read it is to if `is_reverse` is true. (We're re-using a `Position` object here, but the semantics opf this usage are not well documented in the `Position` documentation, and they need to be.) There's also the caveat that if you are working with a graph that doesn't cover the whole chromosome, offset 0 on the path will just be the first base in the region, and not the first base in the chromosome.

For 3, `quality`, when you convert to JSON, comes out base64-encoded. It is a `bytes` array in Protobuf, and each byte is the raw (0 offset) Phred quality for the corresponding base.

For 4, we have some documentation on how Position offsets work:

https://github.com/vgteam/vg/blob/4460c770d66c7dacce356e511d832a31fd6d6cfc/src/vg.proto#L66-L83

Basically, if you are forward, you offset from the start of the node, and if you are reverse, you offset from the end of the node. So the example you gave starts after the leftmost 148 bases on node 8545.

**Comment by derosejf** [comment]
Hi there,

I was hoping for clarification on the construction of a vg from the reference genome.

From my understanding, the nodes and paths in a vg is constructed from a reference genome and a variants file. However, the output files, whether as a vg, GAM or json, work in terms of node ids and offsets instead of linear reference.

Is there any way to recover the original reference / linear coordinates from which the vg nodes were created?

i.e. Is there any ability currently to access the linear coordinates of a particular node (start/end, offset from start of chromosome) or generate an output file containing these coordinates? The json output contains:

`"refpos": [{"offset": 8540756, "name": "chr21"}]`

Which may indicate the linear coordinates.

If there is no method of extracting the coordinates easily, how are the linear coordinates of the reference genome used in the construction of the vg? If they are used for construction of the vg, there is the potential for gathering this information during the construction process for later use, or even extracting it straight from the vg file.

If not, is there a way to infer this information e.g. through post-hoc alignment of the vg graph to a linear reference such as hg38.

This would be very useful for a data visualization that we are building. (Apologies if I missed anything obvious in the the documentation).

**Comment by glennhickey** [comment]
The linear reference is encoded as a paths (one per chrom) in your graph.

To find the position of a node in a particular path, or locate the node at
a given path position you can use `vg find`.

First, make an xg index: `vg index -x graph.xg graph.vg`.

Then for example to find the position of Node N in path R:  `vg find -x
graph.xg -n N -P R`

To see which paths are in your graph: `vg paths -x graph.xg -L`

**Comment by derosejf** [comment]
You mention that to see which paths are in the graph, `vg paths -x graph.xg -L` can be used. This generates nothing to the console when used on my created graph. Is the output supposed to be piped into some kind of output file?

This was tried on a *vg file containing multiple chromosomes, indicating there should be many paths. Any insight you have into this paths command would be greatly appreciated.

**Comment by edawson** [comment]
@derosejf does running `vg paths` on your vg file (i.e. `vg paths -L -v graph.vg`) produce any output?

**Comment by derosejf** [comment]
@edawson Thanks that seemed to fix up the problem.

Just wanted to clarify one more part of the GAM format. When a strand is listed as ```is_reverse = "true"```, this strand is representing a reverse compliment strand, correct? If we were to interpret the node's sequence given this information, is the sequence already represented as the reverse compliment in the GAM or does it need to be inverted?

Further, does this sequence start at its ```refpos``` and continue forward along the reference strand (ending at a higher ```refpos```) or does it walk backwards along the reference strand (ending at a lower ```refpos```)? For example if a reverse sequence started with ```refpos = 100``` and had a sequence length of 10, would it occupy 100-110 or 100-90?

Thanks for you continued help

**Comment by faithokamoto** [comment]
The `sequence` in the GAM is always the read's sequence, so it would need to be reverse complemented to match the nodes in the graph, or visa versa.

Aligning a read or its reverse complement will get the same `refpos` offset value, but the reverse complement one will also have `is_reverse` set to `true` to indicate that the direction you walk is opposite. Your 100-110 option.
